src/ledger/result_writer.zig

```diff
@@ -110,58 +110,143 @@ pub const LedgerResultWriter = struct {
+            const slot, const hash = slot_hash;
+            try setter.addSlotAndHash(slot, hash);
+        }
+
+        try setter.commit();
+    }
+
+    /// Returns a struct which can be used to enact the same operation as `setDuplicateConfirmedSlotsAndHashes`, incrementally.
+    pub fn setDuplicateConfirmedSlotsAndHashesIncremental(
+        self: *Self,
+    ) !SetDuplicateConfirmedSlotsAndHashesIncremental {
+        return .{
+            .result_writer = self,
+            .write_batch = try self.db.initWriteBatch(),
+            .is_committed = false,
+        };
+    }
+    pub const SetDuplicateConfirmedSlotsAndHashesIncremental = struct {
+        result_writer: *Self,
+        write_batch: BlockstoreDB.WriteBatch,
+        is_committed: bool,
+
+        pub fn deinit(self: *SetDuplicateConfirmedSlotsAndHashesIncremental) void {
+            std.debug.assert(self.is_committed);
+            self.write_batch.deinit();
+        }
+
+        pub fn addSlotAndHash(
+            self: *SetDuplicateConfirmedSlotsAndHashesIncremental,
+            slot: Slot,
+            frozen_hash: Hash,
+        ) !void {
+            std.debug.assert(!self.is_committed);
+            const data: FrozenHashVersioned = .{ .current = .{
                 .frozen_hash = frozen_hash,
                 .is_duplicate_confirmed = true,
             } };
-            try write_batch.put(schema.bank_hash, slot, data);
+            try self.write_batch.put(schema.bank_hash, slot, data);
+        }
+
+        pub fn cancel(self: *SetDuplicateConfirmedSlotsAndHashesIncremental) void {
+            std.debug.assert(!self.is_committed);
+            self.is_committed = true;
         }
-        try self.db.commit(&write_batch);
-    }
+
+        pub fn commit(self: *SetDuplicateConfirmedSlotsAndHashesIncremental) !void {
+            std.debug.assert(!self.is_committed);
+            try self.result_writer.db.commit(&self.write_batch);
+            self.is_committed = true;
+        }
+    };
 
     /// agave: set_roots
     pub fn setRoots(self: *Self, rooted_slots: []const Slot) !void {
-        var write_batch = try self.db.initWriteBatch();
-        defer write_batch.deinit();
-        var max_new_rooted_slot: Slot = 0;
-        for (rooted_slots) |slot| {
-            max_new_rooted_slot = @max(max_new_rooted_slot, slot);
-            try write_batch.put(schema.rooted_slots, slot, true);
-        }
+        var setter = try self.setRootsIncremental();
+        defer setter.deinit();
+        errdefer setter.cancel();
+        for (rooted_slots) |rooted_slot| try setter.addRoot(rooted_slot);
+        try setter.commit();
+    }
 
-        try self.db.commit(&write_batch);
-        _ = self.max_root.fetchMax(max_new_rooted_slot, .monotonic);
+    /// Returns a struct which can be used to enact the same operation as `setRoots`, incrementally.
+    pub fn setRootsIncremental(self: *Self) !SetRootsIncremental {
```

Comment:
What about:?

```suggestion
    pub fn setRootsIncremental(self: *LedgerResultWriter) !SetRootsIncremental {
```

Given the recent convention is to use the structs name. It also makes it a tad clearer what `.result_writer` is being set to.

Reply:
f9ee800214ea46aee9349fb789bd70d92659de32